packages/expo-modules-core/ios/Swift/Logging/PersistentFileLog.swift

```diff
@@ -0,0 +1,156 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+import Foundation
+
+public typealias PersistentFileLogFilter = (_: String) -> Bool
+public typealias PersistentFileLogCompletionHandler = (_:Error?) -> Void
+
+/**
+ * A thread-safe class for reading and writing line-separated strings to a flat file.
+ * The main use case is for logging specific errors or events, and ensuring that the
+ * logs persist across application crashes and restarts (for example, OSLogReader can
+ * only read system logs for the current process, and cannot access anything logged
+ * before the current process started).
+ *
+ * All write access to the file goes through asynchronous public methods managed by a
+ * serial dispatch queue.
+ *
+ * The dispatch queue is global, to ensure that multiple instances accessing the same
+ * file will have thread-safe access.
+ *
+ * The only operations supported are
+ * - Read the file (synchronous)
+ * - Append one or more entries to the file
+ * - Filter the file (only retain entries that pass the filter check)
+ * - Clear the file (remove all entries)
+ *
+ */
+public class PersistentFileLog {
+  private static let EXPO_UPDATES_LOG_QUEUE_LABEL = "dev.expo.updates.logging"
+  private static let serialQueue = DispatchQueue(label: EXPO_UPDATES_LOG_QUEUE_LABEL)
+
+  private let category: String
+  private let filePath: String
+
+  public init(category: String) {
+    self.category = category
+    let fileName = "\(PersistentFileLog.EXPO_UPDATES_LOG_QUEUE_LABEL).\(category).txt"
+    self.filePath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first?.appendingPathComponent(fileName).path ?? ""
+  }
+  /**
+   Read entries from log file
+   */
+  public func readEntries() -> [String] {
+    if getFileSize() == 0 {
+      return []
+    }
+    do {
+      return try self.readFileSync()
+    } catch {
+      return []
+    }
+  }
+
+  /**
+   Append entry to the log file
+   Since logging may not require a result handler, the handler parameter is optional
+   If no error handler provided, print error to the console
+   */
+  public func appendEntry(entry: String, _ completionHandler: PersistentFileLogCompletionHandler? = nil) {
+    PersistentFileLog.serialQueue.async {
+      self.ensureFileExists()
+      let size = self.getFileSize()
+      if size == 0 {
+        self.appendTextToFile(text: entry)
+      } else {
+        self.appendTextToFile(text: "\n" + entry)
+      }
```

Comment:
What about 👇 ?
```suggestion
      self.appendTextToFile(text: entry + "\n")
```

Reply:
I would think that this would result in an extra empty line when reading the file, but maybe that's ok. I'll try it.

Reply:
Yes this works, with an adjustment in the reader to handle 0 length lines.